File: 03-week-3--additional-practice/00-bonus-1--balancing-parenetheses/ruby/balancing_parentheses.rb
def balancing_parentheses(string)

  open_par_count = 0
  unmatched_par_count = 0

  string.chars.each do |char|
    if char == '(' 
      open_par_count += 1
      next
    end

    if open_par_count > 0
      open_par_count -= 1
    else
      unmatched_par_count += 1
    end

  end

  open_par_count + unmatched_par_count
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 0"
  puts balancing_parentheses('((c)())')

  puts

  puts "Expecting: 2"
  puts balancing_parentheses('()))')

  puts

  puts "Expecting: 1"
  puts balancing_parentheses(')')

  # Don't forget to add your own!

  puts 'Expecting 3'
  puts balancing_parentheses('((((())))))))')
end

# Please add your pseudocode to this file
# And a written explanation of your solution

File: 03-week-3--additional-practice/00-bonus-1--balancing-parenetheses/javascript/balancing_parentheses.js
function balancingParentheses(string) {
  // type your code here
  chars = string.split('');

  openPars = 0
  missingPairs = 0

  for (let i = 0; i < string.length; ++i ) {

   const char = string[i];

    if ( char === '(' ) {
      ++openPars;
      continue;
    }

    if ( openPars > 0 ) {
      --openPars;
    }
    else {
      ++missingPairs;
    }
  };

  return openPars + missingPairs;

}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 0");
  console.log(balancingParentheses('(()())'));

  console.log("");

  console.log("Expecting: 2");
  console.log(balancingParentheses('()))'));

  console.log("");

  console.log("Expecting: 1");
  console.log(balancingParentheses(')'));
}

module.exports = balancingParentheses;

// Please add your pseudocode to this file
// And a written explanation of your solution
